### PR TITLE
Serial

### PR DESCRIPTION
**Pull request by andrew103**
Adds serial comm processing to the Arduino for both sensor data requests and receiving joint angles. Gives back random sensor data on request.

### CODE REVIEW

File: ard_controller/ard_controller.ino
Comment (goobta) [inline ard_controller/ard_controller.ino:19]: I'm assuming that an analog for this will have to be implemented on the python/ros side?

File: ard_controller/ard_controller.ino
Comment (goobta) [inline ard_controller/ard_controller.ino:199]: Will we ever have a situation when we are asking for a specific sensor reading and not just all of them at once?

File: ard_controller/ard_controller.ino
Comment (goobta) [inline ard_controller/ard_controller.ino:276]: idk how useful this is to implement on the arduino side, but would it be possible to have something like a `MessageBuilder` class? Figure that every single message b/w the arduino & raspi will need to have a preamble as well as a checksum. If we had something like a `QueueMessage()` method or something like that, then `raw_sum` could automatically be tracked.

Then, having a `send()` method is trivial, especially cause the preamble and the checksum will be automatically computed.

**Review by goobta — COMMENTED** [review]
Just a couple of small points. Also, is it possible to write tests for Arduino? I'm assuming not, but there's a lotta logic here so testing it wouldn't be the worst idea in the world.

Not a domain expert on this unfortunately, so gonna hold off and let @mahajanrevant be the final approval on this.

**Comment by andrew103** [comment]
> Just a couple of small points. Also, is it possible to write tests for Arduino? I'm assuming not, but there's a lotta logic here so testing it wouldn't be the worst idea in the world.
> 
> Not a domain expert on this unfortunately, so gonna hold off and let @mahajanrevant be the final approval on this.

Yea unit testing isn't really a thing in Arduino. There are instances of people doing it, but it's not really common practice. Also, there aren't many functions here that actually return anything (for now at least) but instead assign variables, so unit testing would only check the variables which we can just do with a serial print to the console. Also also, the source of the message is on a different platform so you'd basically just be verifying that Arduino can properly assign variables.

File: ard_controller/ard_controller.ino
Comment (andrew103) [reply ard_controller/ard_controller.ino:19]: On the python/ros side, it just has to send the bytes in order, so no need for a full on state machine. You only need a state machine for reading in serial data (which does exist on the python/ros side still for the sensor data service).

File: ard_controller/ard_controller.ino
Comment (andrew103) [reply ard_controller/ard_controller.ino:199]: There might be, but the primary purpose of the service is for the tracker ROS node to keep track of all joints and IMU angles to make sure things are on the right track and to determine when to send the next setpoint. Also, sending all sensor data versus data on specific sensors has a negligible difference in time cost, likely in the order of microseconds or even nanoseconds. Also also, adding the functionality to request specific sensors instead of all of them at once would introduce significant complexity to the state machine to read the serial input since now there would have to be bytes to identify what sensor the proceeding bytes represent instead of relying on the values being sent in a specific order.

File: ard_controller/ard_controller.ino
Comment (andrew103) [reply ard_controller/ard_controller.ino:276]: We could definitely do that, but I would think of that path as optional for the moment. The way I see it, the service response will be the only time the Arduino ever sends data through the serial port (unless we decide to implement a heartbeat for any reason). Since that's the reality, the only reason for creating an external class would be for organization. What would probably be a better idea to do (later on not right now) is to have a class dedicated to serial comms in general that would handle both incoming and outgoing data similar to the serial interface node on the ROS side.

File: ard_controller/ard_controller.ino
Comment (goobta) [reply ard_controller/ard_controller.ino:199]: makes sense! haha that's def not how the current json pipeline works, but that's why we're developing this! :)

File: ard_controller/ard_controller.ino
Comment (goobta) [reply ard_controller/ard_controller.ino:276]: yea also realized that this the arduino and not the ros code--does arduino even have classes lol? yeah, I'd say we can move the message builder to a python package down the line.

**Comment by goobta** [comment]
All of my questions are answered. @mahajanrevant I'll let you approve.

File: ard_controller/ard_controller.ino
Comment (mahajanrevant) [inline ard_controller/ard_controller.ino:127]: Explain this a little bit in the comment?

File: ard_controller/ard_controller.ino
Comment (mahajanrevant) [inline ard_controller/ard_controller.ino:126]: How do we differentiate between front and rear joints?
I think we should have a differentiator in the variable name

File: ard_controller/ard_controller.ino
Comment (mahajanrevant) [inline ard_controller/ard_controller.ino:261]: Why is there a random here? I am assuming for testing?
If that is the case, can you replace it with a method that can have this random implementation?
I don't think it's a good idea to test within the core functionality methods. Let's abstract out the details and test it that way

File: ard_controller/ard_controller.ino
Comment (mahajanrevant) [inline ard_controller/ard_controller.ino:295]: I am assuming this is where we will actuate the joints?

**Review by mahajanrevant — COMMENTED** [review]
Some comments

File: ard_controller/ard_controller.ino
Comment (andrew103) [reply ard_controller/ard_controller.ino:126]: There already is a differentiator. the rear joints are labeled "hip" and "knee" while the front are labeled "shoulder" and "elbow"

File: ard_controller/ard_controller.ino
Comment (andrew103) [reply ard_controller/ard_controller.ino:127]: You want me to elaborate more in the comment?

File: ard_controller/ard_controller.ino
Comment (andrew103) [reply ard_controller/ard_controller.ino:295]: yep

File: ard_controller/ard_controller.ino
Comment (andrew103) [reply ard_controller/ard_controller.ino:261]: Yea it's there as a placeholder until the ODrive is setup with the Arduino. I don't agree with spending the time to abstract something out that's going to be deleted/replaced anyway. It's not like we're going to forget that those are there.

File: ard_controller/ard_controller.ino
Comment (mahajanrevant) [reply ard_controller/ard_controller.ino:126]: Awesome!! Did not catch that

File: ard_controller/ard_controller.ino
Comment (mahajanrevant) [reply ard_controller/ard_controller.ino:127]: No just here!! What does catching the request for sensor data mean?

File: ard_controller/ard_controller.ino
Comment (mahajanrevant) [reply ard_controller/ard_controller.ino:261]: You would need a getAngle method or something like that when you actually put in real code right?
So we could put it now and just add this random implementation inside that.
I'll leave this up to your judgment

**Review by mahajanrevant — COMMENTED** [review]
Just need reply on two more comments

File: ard_controller/ard_controller.ino
Comment (andrew103) [reply ard_controller/ard_controller.ino:261]: For now, since I don't know what the implementation of the ODrive code is gonna look like yet, I'll say let's leave it as is. This is easy to replace.

File: ard_controller/ard_controller.ino
Comment (andrew103) [reply ard_controller/ard_controller.ino:127]: It's basically checking for a specific sequence of bytes that would indicate a request for sensor data. The sequence of bytes would be two 0x01 bytes in succession. The reason for choosing 0x01 is because two 0x01 bytes in succession will never happen in the usual message body of joint angles/sensor data.

Comment (mahajanrevant) [reply ard_controller/ard_controller.ino:127]: Awesome. Thank you

File: ard_controller/ard_controller.ino
Comment (mahajanrevant) [reply ard_controller/ard_controller.ino:261]: Got it

**Review by mahajanrevant — APPROVED** [review]
Good work!! I'll implement the same in ros2sim

**Pull request merged**
